Move DEM examples to documentation

shahabgol opened the pull request:
# Description of the problem

This PR moves the DEM examples to the new documentation.

blaisb commented on doc/source/examples/dem/silo/silo.rst:
```diff
@@ -0,0 +1,197 @@
+
+Running the simulation
+----------------------
+This simulation can be launched by (in parallel mode on 32 processes):
+
+.. code-block:: text
+
+  mpirun -np 32 dem_3d silo_Golshan.prm
+
```
Can you give an estimation of run time?

shahabgol replied:
added.

blaisb reviewed (commented):
Very good work. Some stuff to address, but nothing major. I see tremendous efforts went in there that's for sure!

tonielgeitani commented on doc/source/examples/dem/rotating-drum/rotating-drum.rst:
```diff
@@ -0,0 +1,181 @@
+
+Load-balancing updates the distribution of the subdomains between the processes in parallel simulation to achieve better computational performance (less simulation time). Three load-balancing methods are available in Lethe-DEM: ``once``, ``frequent``, or ``dynamic``. Read `this article <https://www.mdpi.com/2227-9717/10/1/79>`_ for more information about different load-balancing methods and their ...
+
+Selecting ``repartition method = once``, requires defining the step at which the code calls load balancing (``load balance step``). ``Dynamic`` ``repartition method`` requires defining ``load balance frequency``, and in ``dynamic`` ``repartition method``, we should define ``load balance threshold`` and ``dynamic load balance check frequency``. In ``dynamic`` load balancing, the code checks the di ...
```
```suggestion
Selecting ``repartition method = once``, requires defining the step at which the code calls load balancing (``load balance step``). ``Frequent`` ``repartition method`` requires defining ``load balance frequency``, and in ``dynamic`` ``repartition method``, we should define ``load balance threshold`` and ``dynamic load balance check frequency``. In ``dynamic`` load balancing, the code checks the distribution of particles among the processors, every ``dynamic load balance check frequency`` steps, and if
```

tonielgeitani replied:
Isn't the method that uses load balance frequency called frequent and not dynamic?

shahabgol replied:
Absolutely right, I made a mistake here. Corrected.

tonielgeitani commented on doc/source/examples/dem/rotating-drum/rotating-drum.rst:
```diff
@@ -0,0 +1,181 @@
+
+In this subsection, the boundary conditions of the DEM simulation are defined. First of all, the ``number of boundary conditions`` is specified. Then for each boundary condition, its information is defined. There are four boundary types: ``fixed_wall``, ``outlet``, ``rotational`` (around the center), and ``translational``. For ``rotational`` motion, ``rotational speed`` and ``rotational vector``  ...
+
+``fixed_wall`` is a static wall, and particles collide with these static walls upon reaching the wall. The only way to move these walls is to move the entire triangulation. If the ``outlet`` condition is chosen for a boundary, particles can leave the simulation domain via this outlet. Using ``rotational`` and ``translational`` boundary conditions, exerts imaginary rotational and translational vel ...
```
```suggestion
``fixed_wall`` is a static wall, and particles collide with these static walls upon reaching them. The only way to move these walls is to move the entire triangulation. If the ``outlet`` condition is chosen for a boundary, particles can leave the simulation domain via this outlet. Using ``rotational`` or ``translational`` boundary conditions, exerts imaginary rotational and translational velocities to that boundary. In other words, the boundary does not move, but the particles that have collisions with these walls feel a rotational or translational velocity from the wall. This feature is used in the rotating drum example. The boundary id of the ``cylinder`` side wall, defined with deal.ii grid generator is 4. We set the ``rotational speed`` equal to 11.6 rad/s, and the cylinder should rotate around its axis (`x` direction).
```

tonielgeitani commented on doc/source/examples/dem/silo/silo.rst:
```diff
@@ -0,0 +1,197 @@
+Mesh
+~~~~~
+
+Despite previous examples, in this example, we use a mesh generated using Gmsh. The Gmsh (with an extension of .msh) file is located inside the example folder.
```
```suggestion
Contrary to previous examples, in this example, we use a mesh generated using Gmsh. The Gmsh (with an extension of .msh) file is located inside the example folder.
```

tonielgeitani replied:
I don't think the word despite can be used in this context.

shahabgol replied:
Corrected.

shahabgol commented on doc/source/examples/dem/dem.rst:
```diff
@@ -0,0 +1,11 @@
+Discrete Element Method
+****************************
+
+We organize the DEM examples from the simplest (first example) to the most complicated example (last example). Example 1 (packing in circle), shows a two-dimensional simulation of particles packing in a circle. This example is extended in example 2 (packing in ball) to a three-dimensional simulation. Example 3 (rotating drum), shows a simulation in which the particles are packed in a cylinder and ...
```
Done.

shahabgol commented on doc/source/examples/dem/packing-in-circle/packing-in-circle.rst:
```diff
@@ -0,0 +1,194 @@
+Packing in Circle
+==================================
+
+This is the first example of Lethe-DEM. This example introduces the concept of parameter files to parametrize Lethe simulations. It is strongly recommended to visit `DEM parameters <https://lethe-cfd.github.io/lethe/parameters/dem/dem.html>`_ for more detailed information on the concepts and physical meanings of the parameters in Lethe-DEM.
```
Hyperlink from the documentation is used instead.

shahabgol commented on doc/source/examples/dem/packing-in-circle/packing-in-circle.rst:
```diff
@@ -0,0 +1,194 @@
+        set insertion random number seed					= 19
+    end
+
+First, the ``insertion method`` is selected. There are two insertion methods (``uniform`` and ``non_uniform``) in Lethe-DEM. In ``uniform`` insertion the particles are inserted uniformly (without randomness in initial location), while in ``non_uniform``, particles are inserted randomly in the insertion box. ``inserted number of particles at each time step`` specifies the desired number of particl ...
```
Added in a note.

shahabgol commented on doc/source/examples/dem/packing-in-circle/packing-in-circle.rst:
```diff
@@ -0,0 +1,194 @@
+      set integration method				 		= velocity_verlet
+    end
+
+These parameters include ``contact detection method`` and its subsequent information (``dynamic contact search size coefficient`` **or** ``contact detection frequency`` for ``dynamic`` **or** ``constant`` contact detection method), ``neighborhood threshold`` (which defines the contact neighbor list size: ``neighborhood threshold`` * particle diameter), ``particle particle contact force method``,  ...
```
Dynamic contact search only needs one parameter (dynamic contact search size coefficient). Constant contact search also only needs one parameter (contact detection frequency). The theory of constant and dynamic contact search algorithms are added.

shahabgol commented on doc/source/examples/dem/rotating-drum/rotating-drum.rst:
```diff
@@ -0,0 +1,181 @@
+
+Running the simulation
+----------------------
+This simulation can be launched (in parallel mode on 64 processes) by:
+
+.. code-block:: text
+
+  mpirun -np 64 dem_3d rotating_drum.prm
```
I have added a warning to report the approximate simulation time. I will set up a smaller case with a smaller number of particles and add it to this example. but I have to test it on my machine first. So I will do it in another PR.

shahabgol commented on doc/source/examples/dem/rotating-drum-with-post-processing/rotating-drum-with-post-processing.rst:
```diff
@@ -0,0 +1,166 @@
+      set contact detection method 		   		 	= dynamic
+      set dynamic contact search size coefficient	= 0.8
+      set neighborhood threshold				 	= 1.3
+      set load balance method				 		= once
+  	  set load balance step					 		= 150000
+      set particle particle contact force method	= hertz_mindlin_limit_overlap
+      set particle wall contact force method        = nonlinear
+      set integration method				 		= velocity_verlet
+    end
+
+
+Boundary Condition
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. code-block:: text
+
+    subsection DEM boundary conditions
+      set number of boundary conditions         = 1
+        subsection boundary condition 0
+            set boundary id						= 4
+            set type              				= rotational
+            set rotational speed				= 11.6
+            set rotational vector x				= 1
+            set rotational vector y				= 0
+            set rotational vector z				= 0
+        end
+    end
+
+
+Simulation control
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. code-block:: text
+
+    subsection simulation control
+      set time step                 		 = 1e-6
+      set time end       					 = 15
+      set log frequency				         = 1000
+      set output frequency            		 = 1000
+    end
```
done.

shahabgol commented on doc/source/examples/dem/rotating-drum-with-post-processing/rotating-drum-with-post-processing.rst:
```diff
@@ -0,0 +1,166 @@
+    end
+
+
+Running the simulation
+----------------------
+This simulation can be launched (in parallel mode on 64 processes) by:
+
+.. code-block:: text
+
+  mpirun -np 64 dem_3d rotating_drum_with_post_processing.prm
```
done.

shahabgol commented on doc/source/examples/dem/packing-in-ball/packing-in-ball.rst:
```diff
@@ -0,0 +1,153 @@
+  mpirun -np 8 dem_3d packing_in_ball.prm
+
+.. note::
+	The parallel simulations are generally faster than simulations on a single process. However, to leverage the full performance of a parallel simulation, it should be performed with a load-balancing strategy throughout the simulation. Load-balancing is explained in the next example.
```
Added as a hyperlink.

shahabgol commented:
@blaisb @tonielgeitani I have addressed your comments in the last commit. Only one comment remains on designing a new example for rotating drum with a smaller number of particles. I will do it in a separate PR, as I need to have access to my workstation. :)
